core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -234,51 +232,50 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         }
 
         @Override
-        public String describeTasks(List<ShardRoutingEntry> tasks) {
-            return tasks.stream().map(entry -> entry.getShardRouting().toString()).reduce((s1, s2) -> s1 + ", " + s2).orElse("");
-        }
-
-        @Override
         public BatchResult<ShardRoutingEntry> execute(ClusterState currentState, List<ShardRoutingEntry> tasks) throws Exception {
             BatchResult.Builder<ShardRoutingEntry> batchResultBuilder = BatchResult.builder();
+            List<ShardRoutingEntry> validTasks = new ArrayList<>();
+            List<FailedRerouteAllocation.FailedShard> validShardRoutings = new ArrayList<>();
 
-            // partition tasks into those that correspond to shards
-            // that exist versus do not exist
-            Map<ValidationResult, List<ShardRoutingEntry>> partition =
-                tasks.stream().collect(Collectors.groupingBy(task -> validateTask(currentState, task)));
+            for (ShardRoutingEntry task : tasks) {
+                IndexMetaData indexMetaData = currentState.metaData().index(task.shardId.getIndex());
+                if (indexMetaData == null) {
+                    // tasks that correspond to non-existent shards are marked as successful
+                    batchResultBuilder.success(task);
+                } else {
+                    // non-local requests
+                    if (task.primaryTerm > 0 && task.primaryTerm != indexMetaData.primaryTerm(task.shardId.id())) {
+                        batchResultBuilder.failure(task, new NoLongerPrimaryShardException(
+                            task.shardId,
+                            "primary term [" + task.primaryTerm + "] did not match current primary term [" +
+                                indexMetaData.primaryTerm(task.shardId.id()) + "]"));
+                        continue;
+                    }
 
-            // tasks that correspond to non-existent shards are marked
-            // as successful
-            batchResultBuilder.successes(partition.getOrDefault(ValidationResult.SHARD_MISSING, Collections.emptyList()));
+                    ShardRouting matched = currentState.getRoutingNodes().getByAllocationId(task.shardId, task.allocationId);
```

Comment:
why is this going through RoutingNodes ? isn't routing table more natural (and doesn't require building the routing nodes variant)?